Rewrite editor-preview package as extensions of editor classes

colin-grant-work opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->

This PR is a more-or-less wholesale rewrite of the `editor-preview` package to extend and override the functionality of the `EditorManager`, `EditorWidgetFactory`, and `EditorWidget` rather than using a wrapper around an `EditorWidget`.

 - Fixes #3634 by making the preview an immediate child of the shell
 - Fixes #4061 by using a normal editor widget
 - Fixes #4708 by reducing the complexity of the docking and widget-tracking logic
 - Fixes #6784, fixes #8511, fixes #9485 by implementing better checks for a widget's prior existence and by using a single opening routine for all editors, rather than having two competing openers for previews and regular editor widgets.

These issues aren't directly addressed by these changes, but I can't reproduce them with this code:
 - Closes #4903
 - Closes #7078

#### Regressions:
- ~~double-clicking the title of a preview editor will not pin that editor.~~
> I think the best way to handle this problem is what I've outlined in #9511, but in the meantime, I've achieved this by using a listener on the `document`.

#### Breaking Changes:
- ~~Change of various return types in `WidgetOpenHandler` and derived classes from `Promise<T | undefined>` to `MaybePromise<T> | undefined` to facilitate synchronous checks for widget existence.~~
> I've removed this breakage by adding methods to the `WidgetManager` and `WidgetOpenHandler` that can check for a widget's (pending) existence synchronously rather than modifying existing methods.


#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->

##### Basics
1. Makes sure that the preference `editor.enablePreview` is `true` and run the `Restore workbench layout` command.
2. Single-click various files in the navigator.
3. [ ] One editor should be created, and it should be replaced with each new file opened.
4. Double click the file you have open in the tree.
5. [ ] That editor should switch to non-preview mode.
6. Single click another file.
7. [ ] A new editor preview should open up.
8. Single click the file in the pinned (non-preview) editor.
9. [ ] The non-preview editor should be revealed, and no new editor should be opened.
10. Focus the current preview editor and start typing.
11. [ ] It should turn into a normal editor.

##### Fixes
 - Renaming
   1. Open a preview, then rename the file.
   2. [ ] The preview with the old name should be replaced with a new one with the new name.
   3. [ ] Single-clicking a different file should replace the preview.
 - Navigating
   1. Open an editor preview in a source code file.
   2. Use context menu 'Go to references' or <kbd>Shift + F12</kbd> to view a references peek widget
   3. [ ] Clicking on the destinations should work fine.
 - Dragging
   1. Open a preview.
   2. Drag it into a side panel
   3. [ ] You should still see the editor.
   4. Drag it back
   5. [ ] It should be a non-preview
 - Rapid opening
   1. Follow the reproduction instructions in #9485 
   2. [ ] You should never wind up with two editors.

##### Others
 - [ ] Using <kbd>ctrl+w</kbd> (Electron) / <kbd>alt+w</kbd> (Browser) to close an editor preview should work.
 - [ ] Opening a preview editor for e.g. a Markdown file and then using the `Open preview` should work as expected.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)


Signed-off-by: Colin Grant <colin.grant@ericsson.com>

kenneth-marut-work reviewed (commented):
This is working much better and it seems to have fixed the issues mentioned above. On a first pass running through the functionality, I only noticed two small behavioral issues related to pinning:

1. Splitting the editor view in the main area
- Open 2 files (one being a preview)
- Drag the preview to the right side of the MAIN area to split the screen
- The preview does not pin

2. Pinning an editor when it does not exist in the file tree
- Open a file using the steps in #9485 with a file that does not exist in your current workspace
- Observe that the file is opened as a preview widget, but it is not possible to pin it easily (now that the double clicking does not pin)
- In VSCode there is a context command for "Keep Editor Open" which might be a useful approach

kenneth-marut-work commented on packages/editor-preview/src/browser/editor-preview-manager.ts:
```diff
@@ -14,151 +14,104 @@
-
-    @inject(EditorManager)
-    protected readonly editorManager: EditorManager;
+    @inject(WidgetManager) protected readonly widgetManager: WidgetManager;
```
Minor: This exists on the WidgetOpenHandler already

colin-grant-work replied:
Good point - I've deleted it here.

colin-grant-work commented:
@kenneth-marut-work, thanks for taking a look. I've made some changes to address your comments
 - changed the drag logic to track the widgets tabbar rather than its parent to cover dragging inside the `main` area.
 - added context menu commands and a keyboard shortcut parallel to VSCode's 'Keep Open' command.
 - brought back the pin-on-double-click logic with a listener on the `document`.

kenneth-marut-work commented on packages/editor-preview/src/browser/editor-preview-widget.ts:
```diff
@@ -15,22 +15,34 @@
         return this._isPreview;
     }
 
+    constructor(
```
Might be an appropriate time to make this and the EditorWidget injectable

kenneth-marut-work commented:
Verified that the new context menu pin option and double clicking work 👍

colin-grant-work commented on packages/core/src/browser/keybinding.ts:
```diff
@@ -245,6 +245,22 @@ export class KeybindingRegistry {
+     * Ensures that keybindings are inserted in order of increasing length of binding to ensure that if a
+     * user triggers a short keybinding (e.g. ctrl+k), the UI won't wait for a longer one (e.g. ctrl+k enter)
+     */
+    protected insertBindingIntoScope(item: common.Keybinding & { scope: KeybindingScope; }, scope: KeybindingScope): void {
+        const scopedKeymap = this.keymaps[scope];
+        const getNumberOfKeystrokes = (binding: common.Keybinding): number => (binding.keybinding.trim().match(/\s/)?.length ?? 0) + 1;
+        const numberOfKeystrokesInBinding = getNumberOfKeystrokes(item);
+        const indexOfFirstItemWithEqualStrokes = scopedKeymap.findIndex(existingBinding => getNumberOfKeystrokes(existingBinding) === numberOfKeystrokesInBinding);
+        if (indexOfFirstItemWithEqualStrokes > -1) {
+            scopedKeymap.splice(indexOfFirstItemWithEqualStrokes, 0, item);
+        } else {
+            scopedKeymap.push(item);
+        }
+    }
+
```
This addition is necessary to ensure that this test passes:  
https://github.com/eclipse-theia/theia/blob/8ea255a4021393eba96a0da0acae3b1aa8f98ad6/examples/api-tests/src/keybindings.spec.js#L48-L63

Previously, it passed by coincidence - the relevant competition to the clear terminal command happened to be registered in an order that allowed the test to pass. 

FWIW, the assertion made in [this test](https://github.com/eclipse-theia/theia/blob/8ea255a4021393eba96a0da0acae3b1aa8f98ad6/examples/api-tests/src/keybindings.spec.js#L96-L115) that priority should depend on insertion order is pretty fragile, and seems designed to cover edge cases from the [`terminal` package](https://github.com/eclipse-theia/theia/blob/8ea255a4021393eba96a0da0acae3b1aa8f98ad6/packages/terminal/src/browser/terminal-frontend-contribution.ts#L430-L505) where we register a bunch of non-commands for <kbd>ctrl+\<key\></kbd> and <kbd>alt+\<key\></kbd> to get around bindings that already exist in the packages we use for terminals, but then *also* register some commands for specific combinations that we want to take priority over the dummy commands. There's got to be a better way to handle those requirements.

vince-fugnitto commented on packages/editor-preview/src/browser/editor-preview-contribution.ts:
```diff
@@ -0,0 +1,71 @@
+import { EditorPreviewWidget } from './editor-preview-widget';
+
+export namespace EditorPreviewCommands {
+    export const PIN_PREVIEW_COMMAND: Command = {
+        id: 'workbench.action.keepEditor',
+        label: 'Keep Open'
+    };
```
```suggestion
    export const PIN_PREVIEW_COMMAND: Command = {
        id: 'workbench.action.keepEditor',
        category: 'View',
        label: 'Keep Editor'
    };
```

colin-grant-work replied:
👍 Done.

vince-fugnitto commented on packages/editor-preview/src/browser/editor-preview-manager.ts:
```diff
@@ -1,5 +1,5 @@
 /********************************************************************************
- * Copyright (C) 2018 Google and others.
+ * Copyright (C) 2021 Ericsson and others.
```
We should keep the original license (generally we do not update them - they reflect the original author of the file).

colin-grant-work replied:
This is a slightly tricky case - I created whole new files with separate names (and new headers) while I was working on the code, then replaced the old files when I was done, so there is very little lineal connection between the file by this name on `master` and the version here - the diff shows almost only brackets 'carried over'. Since this is a totally breaking change (if you were importing anything from any of these files, it no longer does what you thought), there's also a case for using different file names so that's impossible to pretend that they still do what they used to. I'm not an expert on the ins and outs of the license headers, though, so I defer to the judgment of anyone better informed.

colin-grant-work replied:
Changed to `2018-2021 Google and others`

vince-fugnitto commented on packages/editor-preview/src/browser/editor-preview-widget-factory.ts:
```diff
@@ -0,0 +1,45 @@
+import { EditorPreviewWidget } from './editor-preview-widget';
+import { NavigatableWidgetOptions } from '@theia/core/lib/browser';
+
+export interface EditorPreviewOptions extends NavigatableWidgetOptions {
```
Do we need both interfaces?
https://github.com/eclipse-theia/theia/pull/9518/files#diff-075aefa08a4fb3b0224a5bcc327bf6ed6f9b9aa1d8ff0fa9ad159b1a89406b3cR27

colin-grant-work replied:
Likely not. Kept the one in the factor, removed it here. 😅

vince-fugnitto commented on packages/editor-preview/src/browser/editor-preview-widget.ts:
```diff
@@ -1,5 +1,5 @@
 /********************************************************************************
- * Copyright (C) 2018 Google and others.
+ * Copyright (C) 2021 Ericsson and others.
```
Same as earlier comment.

colin-grant-work replied:
changed to `2018-2021 Google and others`

vince-fugnitto reviewed (approved):
I verified the following functionality :+1: 

**Preference `editor.enablePreview`**:

- [x] the preference is enabled by default
- [x] when _true_ editors are opened as a preview and replaced
- [x] when _false_ editors are opened normally and not replaced
- [x] when toggling from _true_ to _false_, the open _preview_ editor is transformed into a normal editor 

**Application**

- [x] a _preview_ editor is transformed into a normal editor when it is modified by a user
- [x] a _single-click_ in the explorer opens a _preview_ editor when the preference is on
- [x] a _single-click_ in the explorer opens a normal editor when the preference is off
- [x] a double-click in the explorer opens a normal editor (preference on)
- [x] renaming a _preview_ editor works correctly - the editor remains a _preview_
- [x] dragging _preview_ editors works correctly (sidepanel to main-area and so-on)
- [x] keybinding successfully closes a _preview_ editor
- [x] splitting a _preview_ editor works
- [x] reloading the application with _preview_ editors works 
- [x] performing toolbar actions (ex: 'markdown: open preview') should work for _preview_ editors

colin-grant-work commented:
@caseyflynn-google, this may be of interest to you. I plan to merge on Monday or Tuesday, if there are no objections.

kenneth-marut-work commented on packages/editor-preview/src/browser/editor-preview-contribution.ts:
```diff
@@ -0,0 +1,72 @@
+    export const PIN_PREVIEW_COMMAND: Command = {
+        id: 'workbench.action.keepEditor',
+        category: 'View',
+        label: 'Keep Editor',
```
minor: in VSCode, this label that appears in the context menu is 'Keep Open'

vince-fugnitto replied:
> minor: in VSCode, this label that appears in the context menu is 'Keep Open'

Then we would need to update the label at: https://github.com/eclipse-theia/theia/pull/9518/files#diff-14f606c80287893ed697301e733999e7682fe47f8680bb99d05feeb40f446d16R64.

colin-grant-work replied:
I've modified the label below.

kenneth-marut-work commented:
Found one small issue, right clicking a file and using the "Open With" menu option reveals 2 "Code Editor" openers

![image](https://user-images.githubusercontent.com/62660714/120663961-24dfb300-c450-11eb-8360-079b615033c0.png)

colin-grant-work commented:
> Found one small issue, right clicking a file and using the "Open With" menu option reveals 2 "Code Editor" openers

@kenneth-marut-work, good find. Since `WidgetOpenHandler` was already bound to the `EditorManager`, I didn't need to bind it to the `EditorPreviewManager` again. I've removed that, and now there should be 1 (or 0) 'Open With' entries.

caseyflynn-google reviewed (approved):
Awesome work! Verified all items listed in the testing section:

> #### How to test
> ##### Basics
> 1. Makes sure that the preference `editor.enablePreview` is `true` and run the `Restore workbench layout` command.
> 2. Single-click various files in the navigator.
> 3. [x]  One editor should be created, and it should be replaced with each new file opened.
> 4. Double click the file you have open in the tree.
> 5. [x]  That editor should switch to non-preview mode.
> 6. Single click another file.
> 7. [x]  A new editor preview should open up.
> 8. Single click the file in the pinned (non-preview) editor.
> 9. [x]  The non-preview editor should be revealed, and no new editor should be opened.
> 10. Focus the current preview editor and start typing.
> 11. [x]  It should turn into a normal editor.
> 
> ##### Fixes
> * Renaming
>   
>   1. Open a preview, then rename the file.
>   2. [x]  The preview with the old name should be replaced with a new one with the new name.
>   3. [x]  Single-clicking a different file should replace the preview.
> * Navigating
>   
>   1. Open an editor preview in a source code file.
>   2. Use context menu 'Go to references' or Shift + F12 to view a references peek widget
>   3. [x]  Clicking on the destinations should work fine.
> * Dragging
>   
>   1. Open a preview.
>   2. Drag it into a side panel
>   3. [x]  You should still see the editor.
>   4. Drag it back
>   5. [x]  It should be a non-preview
> * Rapid opening
>   
>   1. Follow the reproduction instructions in #9485
>   2. [x]  You should never wind up with two editors.
> 
> ##### Others
> * [x]  Using ctrl+w (Electron) / alt+w (Browser) to close an editor preview should work.
> * [x]  Opening a preview editor for e.g. a Markdown file and then using the `Open preview` should work as expected.
> 

Additional checks:
- [x] All context menu items in editor preview tab work (Close, Close Others, Toggle Maximized etc)
- [x] Language server features work in editor preview
- [x] Editor reviews are properly revived (only tested in browser)
